stdlib/public/runtime/Backtrace.cpp

```diff
@@ -789,20 +843,100 @@ namespace backtrace {
 /// @param mangledName is the symbol name to be tested.
 ///
 /// @returns `true` if `mangledName` represents a thunk function.
-SWIFT_RUNTIME_STDLIB_SPI SWIFT_CC(swift) bool
-_swift_isThunkFunction(const char *mangledName) {
+SWIFT_RUNTIME_STDLIB_SPI bool
+_swift_backtrace_isThunkFunction(const char *mangledName) {
   swift::Demangle::Context ctx;
 
   return ctx.isThunkSymbol(mangledName);
 }
 
+// Try to demangle a symbol.
+SWIFT_RUNTIME_STDLIB_SPI char *
+_swift_backtrace_demangle(const char *mangledName,
+                          size_t mangledNameLength,
+                          char *outputBuffer,
+                          size_t *outputBufferSize,
+                          int *status) {
+  llvm::StringRef name = llvm::StringRef(mangledName, mangledNameLength);
+
+  // You must provide buffer size if you're providing your own output buffer
+  if (outputBuffer && !outputBufferSize) {
+    return nullptr;
+  }
+
+  if (Demangle::isSwiftSymbol(name)) {
+    // This is a Swift mangling
+    auto options = DemangleOptions::SimplifiedUIDemangleOptions();
+    auto result = Demangle::demangleSymbolAsString(name, options);
+    size_t bufferSize;
+
+    if (outputBufferSize) {
+      bufferSize = *outputBufferSize;
+      *outputBufferSize = result.length() + 1;
+    }
+
+    if (outputBuffer == nullptr) {
+      outputBuffer = (char *)::malloc(result.length() + 1);
```

Comment:
I think this needs a pointer check to make sure it's not a `nullptr`.

Reply:
I'm normally of the "don't bother checking if `malloc()` returned `NULL`" school of thought, especially on Linux where the allocator overcommits (at least, unless you force that off) and where _actually_ running out of memory might cause the OOM killer to arbitrarily victimise something anyway.

I guess we _could_ check and return _nullptr_, but we're probably in such a mess already if that happens that there's little hope of any good outcome.

Reply:
Isn't it set `*status` to `-1` in that case?

Reply:
Ah, I see what you're talking about. Maybe I should just remove `status`, since we don't use it anyway.